tests/unit/test_charm.py

```diff
@@ -12,13 +12,81 @@ def setUp(self) -> None:
         self.harness.begin()
         self.maxDiff = None
 
+    @patch("ops.framework.EventBase.defer")
+    def test_check_etcd_fail(self, defer):
+        self.harness.charm.on.check_etcd.emit()
+        defer.assert_called()
+
+    @patch("charm.SlurmdCharm._on_slurmctld_started")
+    @patch("json.loads", lambda _: ["test"])
+    @patch("charm.SlurmdCharm.hostname", new_callable=PropertyMock(return_value="test"))
+    @patch("omnietcd3.Etcd3AuthClient.get")
+    @patch("charm.SlurmdCharm.etcd_ca_cert", new_callable=PropertyMock(return_value=""))
+    @patch("charm.SlurmdCharm.etcd_use_tls", new_callable=PropertyMock(return_value=False))
+    @patch("ops.framework.EventBase.defer")
+    def test_check_etcd_success(self, defer, *_):
+        self.harness.charm.on.check_etcd.emit()
+        defer.assert_not_called()
+
+    @patch("ops.framework.EventBase.defer")
+    def test_config_changed_fail(self, defer):
+        self.harness.set_leader(True)
+        self.harness.charm.on.config_changed.emit()
+        defer.assert_called()
+
+    @patch("ops.framework.EventBase.defer")
+    def test_config_changed_success(self, defer):
+        self.harness.charm.on.config_changed.emit()
+        defer.assert_not_called()
+    
+    @patch("ops.framework.EventBase.defer")
+    def test_install_fail(self, defer):
+        self.harness.charm.on.install.emit()
+        self.assertFalse(self.harness.charm._stored.slurm_installed)
+        defer.assert_called()
+
+    @patch("slurm_ops_manager.SlurmManager.install")
+    @patch("pathlib.Path.read_text")
+    @patch("ops.model.Unit.set_workload_version")
+    @patch("ops.model.Resources.fetch")
+    @patch("ops.framework.EventBase.defer")
+    def test_install_success(self, defer, *_):
+        self.harness.charm.on.install.emit()
+        self.assertTrue(self.harness.charm._stored.slurm_installed)
+        defer.assert_not_called()
```

Comment:
Do we want to mock slurmd being installed? To me this method's success is hinged on the call out to  SlurmManager's install method actually working. I think we should not patch this out since it is the install method; it will not succeed in production if `SlurmManager.install` fails.

Also, should we not patch out `pathlib.Path.read_text` and `ops.model.Unit.set_work_load` version for install? I have install intentionally failing in slurmctld and slurmdbd to point out that the install will not work unless the version file is present. Opens the discussion for refectoring how the slurm charms' version is set.

Reply:
Hmm, I thought about this for a bit and I think I do want to mock it being installed. Reason being this keeps the unit tests simple and only testing charm results like "if I run x what is the resultant status/_stored state etc" or if a defer was called. 

For testing the actual install that will go under `functional/test_slurmd.py` where in my `setUp` I'll install `slurmd` and then have a bunch of test cases to test the functionality/capabilities of slurmd for normal/expected operations. I agree about the failing test and how a failing test is like a code smell of something in need of a factor however I do no want to have a failing test it can send the wrong message or be confusing. Open to discuss this more and the refactor path forward

Reply:
Your reasoning for mocking the install of slurmd makes sense to me now that you have explained it. I agree that when testing `charm.py` you are looking to make sure that the proper deferral is made and/or proper status message is shown.

That said, I still think that you should not mock `pathlib.Path.read_text` because I think that presents a real problem to proper event management/correct status message being shown.

Reply:
Looks good :smile:. I see that you are still patching out version, but your reasoning makes sense to me. Just make a mental note that we need to fix the version file issue or we are going to have a lot of unhappy campers.